[DOCS] Update note to add further details on current status of YUL

ChrisChinchilla opened the pull request:
### Description

Closes https://github.com/ethereum/solidity/issues/6268 I think this addresses the issue, but I'm only starting to understand YUL, so likely missed or misunderstood something.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

Commit:
Update note to add further details on current status of YUL

Clarify further

Changes from reviews

Grammar polish

Changes from review

Update docs/yul.rst

Co-Authored-By: ChrisChinchilla <chriswhward@gmail.com>

Update docs/yul.rst

Co-Authored-By: ChrisChinchilla <chriswhward@gmail.com>

Polish suggestions

Update docs/yul.rst

Co-Authored-By: ChrisChinchilla <chriswhward@gmail.com>

chriseth commented on docs/yul.rst:
```diff
@@ -15,13 +15,16 @@ It can already be used for "inline assembly" inside Solidity and
+.. warning::
 
-    Note that the flavour used for "inline assembly" does not have types
+    Yul does not yet support types
```
The thing is that it is still a different language. So "strict assembly" is implemented (and that's also what it is called when using the commandline interface), but "Yul" as described below is not.

ChrisChinchilla commented:
OK @chriseth I attempted to understand what you said, and spoke to some of the team for further clarification, I hope this is getting closer…

bshastry commented on docs/yul.rst:
```diff
@@ -9,18 +9,19 @@ Yul
-.. note::
+Yul is implemented when using the compiler in strict assembly mode, which you can use
+by adding the ``--strict-assembly`` and ``--optimize`` arguments when using ``solc``.
+
```
I was of the impression that Yul is not yet available to the public for use. Perhaps Chris meant something like "to use the present incarnation of the optimizer, use `solc --strict-assembly --optimize.`

I realize this is confusing terminology which I also had trouble getting my head around :) FWIW, I'm adding my notes below
  - `yul` is the future optimizer (and also a namesake intermediate langugage)
    - the `yul` optimizer supports two intermediate language dialects, namely: `strict assembly` and `yul`
    - `yul` (the intermediate language) is not yet supported but `strict assembly` is
    - people can use the `yul` optimizer by passing the `--strict-assembly --optimize` flag to `solc`. Essentially, they get the benefits of a modern optimizer (`yul`) without having to wait for the release of the namesake intermediate language.

ChrisChinchilla replied:
I am also somewhat confused as I haven't really been exposed to yul, but this all helps I think. I'll take another pass soon. 👍

chriseth replied:
This is almost correct. We don't really use "Yul" as a term for the optimizer, we usually say "yul optimizer" if we talk about it. Yul is the (intermediate) language. "Using yul" in my mind is taking yul source code, potentially running the optimizer and then generating bytecode output. This bytecode output is not yet implemented for the "strict yul" flavour that has types and a different set of built-in functions. It is fully implemented for the "strict assmebly" flavour which does not have types and uses the EVM instructions as built-in functions.

bshastry commented on docs/yul.rst:
```diff
@@ -9,18 +9,19 @@ Yul
-    to the EVM opcodes. Please resort to the inline assembly documentation
-    for details.
+    Yul is in active development, and as such does not yet support types (everything is ``u256``),
+    and the built-in functions are identical to the EVM opcodes. Refer to the :ref:`inline assembly documentation <inline-assembly>` for details.
```
Yul is indeed in active development, however it does support types. I suggest retaining the note, and adding a sentence to the effect of "Yul is not supported yet because it is under active development."

ChrisChinchilla replied:
OK, this I am confused by, as the existing text said it doesn't support types and I'm sure that's what @chriseth also said as a reason to create the issue. So I need a little clarity there :)

chriseth commented on docs/yul.rst:
```diff
@@ -7,21 +7,19 @@ Yul
+    ``--strict-assembly`` and ``--optimize`` arguments) flavour which does not have types (everything is ``u256``)
+    and uses :ref:`EVM opcodes <opcodes>` as built-in functions.
+
+    It is not yet implemented for the "strict yul" flavour that has types and a different set of built-in functions.
```
`The "strict yul" flavour is not yet implemented..'?

ChrisChinchilla commented:
@chriseth @bshastry Another attempt to bring in all suggestions, and there was another (and related) merged PR to this file that I also brought into this change to help the flow.

bshastry commented on docs/yul.rst:
```diff
@@ -7,15 +7,13 @@ Yul
+platforms. It can already be used for "inline assembly" inside Solidity and future versions of the Solidity compiler
+will use Yul as an intermediate language. It should also be easy to build high-level optimizer stages for Yul.
+
+With the "inline assembly" flavour, Yul can be used as a language setting
```
Is "inline assembly" a flavour as well or is it just a way for end-users to write code? From what I gather, "strict assembly" and "typed strict assembly" are the flavours. "inline assembly" is simply the use of assembly in Solidity contracts.

ekpyron commented on docs/yul.rst:
```diff
@@ -7,15 +7,13 @@ Yul
-language. It should also be easy to build high-level optimizer stages for Yul.
-
-In its flavour of inline-assembly, Yul can be used as a language setting
+compile bytecode output for different backends.
```
Is it really correct to say ``Yul can compile bytecode output``? Shouldn't it rather be ``Yul can *be compiled to* bytecode``?

ekpyron commented on docs/yul.rst:
```diff
@@ -7,15 +7,13 @@ Yul
+
+Support for EVM 1.0, EVM 1.5 and eWASM is planned, and it is designed to be a usable common denominator of all three
+platforms. It can already be used for "inline assembly" inside Solidity and future versions of the Solidity compiler
+will use Yul as an intermediate language. It should also be easy to build high-level optimizer stages for Yul.
```
```suggestion
will use Yul as an intermediate language. Yul is a good target for high-level optimization stages (that can benefit all target platforms equally).
```
?
Maybe even further ``An experimental Yul optimizer is already part of the Solidity compiler.``?

ekpyron commented on docs/yul.rst:
```diff
@@ -7,15 +7,13 @@ Yul
+compile bytecode output for different backends.
+
+Support for EVM 1.0, EVM 1.5 and eWASM is planned, and it is designed to be a usable common denominator of all three
+platforms. It can already be used for "inline assembly" inside Solidity and future versions of the Solidity compiler
```
Maybe we should remove this sentence about "inline assembly" here (continuing directly with Yul as a good language for optimization and an experimental Yul optimizer being part of the Solidity compiler). Then after this we could have a new paragraph explaining the whole flavor thing:

``Yul comes in two flavours, namely untyped Yul and typed Yul.`` (naming up for discussion, but I'd prefer this over "strict" and "non-strict" to avoid confusion) ``While future versions of Solidity will use (typed) Yul as intermediate language, untyped Yul can already be used in Solidity as "strict inline assembly". This mode for inline assembly can be enabled for the command line interface using the --strict-assembly parameter. Furthermore, untyped Yul can be used as language setting for the standard-json interface.``

ekpyron commented on docs/yul.rst:
```diff
@@ -29,15 +27,14 @@ for the :ref:`standard-json interface <compiler-api>`:
-.. note::
+.. warning::
+
+    Yul is in active development and its bytecode output is fully implemented for the "strict assembly" flavour which
```
```suggestion
    Yul is in active development and bytecode generation is fully implemented only for untyped Yul (which does not have types; everything is u256) and with EVM 1.0 as target (EVM opcodes are used as built-in functions).
    Typed Yul will use a different set of built-in functions and is not yet implemented.
```

ekpyron commented:
Some suggestions from my side - I hope this isn't getting horribly confusing now :-).

ChrisChinchilla commented:
Implemented some suggestions from @ekpyron and polished.

chriseth commented on docs/yul.rst:
```diff
@@ -6,16 +6,14 @@ Yul
-
-In its flavour of inline-assembly, Yul can be used as a language setting
+Yul (previously also called JULIA or IULIA) is an intermediate language that can be
+compiled to bytecode output for different backends.
```
```suggestion
compiled to bytecode for different backends.
```

chriseth commented:
Looks good now! Please squash.

stackenbotten commented:
There was an error when running `test_check_style` for commit `1d6b2b11d1f86a54fea7f13388048f0530c73e07`:
```
Coding style error:
 libyul/backends/evm/EVMCodeTransform.cpp:84: [=](Scope::Variable const &_var) {
 libyul/backends/evm/EVMCodeTransform.cpp:85: ++m_context.variableReferences[&_var];
 libyul/backends/evm/EVMCodeTransform.cpp:86: },
 libyul/backends/evm/EVMCodeTransform.cpp:87: [=](Scope::Label const &) {},
 libyul/backends/evm/EVMCodeTransform.cpp:88: [=](Scope::Function const &) {}));
 libyul/backends/evm/EVMCodeTransform.cpp:298: [=](Scope::Variable &) { solAssert(false, "Expected function name."); },
 libyul/backends/evm/EVMCodeTransform.cpp:299: [=](Scope::Label &) { solAssert(false, "Expected function name."); },
 libyul/backends/evm/EVMCodeTransform.cpp:300: [&](Scope::Function &_function) { function = &_function; })),
 libyul/backends/evm/EVMCodeTransform.cpp:301: "Function name not found.");
 libyul/backends/evm/EVMCodeTransform.cpp:357: [=](Scope::Variable &_var) {
 libyul/backends/evm/EVMCodeTransform.cpp:360: if (int heightDiff = variableHeightDiff(_var, _identifier.name, false))
 libyul/backends/evm/EVMCodeTransform.cpp:361: m_assembly.appendInstruction(dev::eth::dupInstruction(heightDiff));
 libyul/backends/evm/EVMCodeTransform.cpp:362: else
 libyul/backends/evm/EVMCodeTransform.cpp:364: m_assembly.appendConstant(u256(0));
 libyul/backends/evm/EVMCodeTransform.cpp:365: decreaseReference(_identifier.name, _var);
 libyul/backends/evm/EVMCodeTransform.cpp:366: },
 libyul/backends/evm/EVMCodeTransform.cpp:367: [=](Scope::Label &_label) {
 libyul/backends/evm/EVMCodeTransform.cpp:368: m_assembly.appendLabelReference(labelID(_label));
 libyul/backends/evm/EVMCodeTransform.cpp:369: },
 libyul/backends/evm/EVMCodeTransform.cpp:370: [=](Scope::Function &) {
 libyul/backends/evm/EVMCodeTransform.cpp:371: solAssert(false, "Function not removed during desugaring.");
 libyul/backends/evm/EVMCodeTransform.cpp:372: })))
 libyul/backends/evm/EVMCodeTransform.cpp:672: [=](Scope::Variable &) { solAssert(false, "Expected label"); },
 libyul/backends/evm/EVMCodeTransform.cpp:673: [&](Scope::Label &_label) {
 libyul/backends/evm/EVMCodeTransform.cpp:674: label = labelID(_label);
 libyul/backends/evm/EVMCodeTransform.cpp:675: },
 libyul/backends/evm/EVMCodeTransform.cpp:676: [=](Scope::Function &) { solAssert(false, "Expected label"); })))
 libyul/backends/evm/EVMCodeTransform.cpp:36:void VariableReferenceCounter::operator()(Identifier const &_identifier)
 libyul/backends/evm/EVMCodeTransform.cpp:41:void VariableReferenceCounter::operator()(FunctionDefinition const &_function)
 libyul/backends/evm/EVMCodeTransform.cpp:49: for (auto const &v : _function.returnVariables)
 libyul/backends/evm/EVMCodeTransform.cpp:57:void VariableReferenceCounter::operator()(ForLoop const &_forLoop)
 libyul/backends/evm/EVMCodeTransform.cpp:71:void VariableReferenceCounter::operator()(Block const &_block)
 libyul/backends/evm/EVMCodeTransform.cpp:84: [=](Scope::Variable const &_var) {
 libyul/backends/evm/EVMCodeTransform.cpp:92: AbstractAssembly &_assembly,
 libyul/backends/evm/EVMCodeTransform.cpp:93: AsmAnalysisInfo &_analysisInfo,
 libyul/backends/evm/EVMCodeTransform.cpp:94: Block const &_block,
 libyul/backends/evm/EVMCodeTransform.cpp:96: EVMDialect const &_dialect,
 libyul/backends/evm/EVMCodeTransform.cpp:98: ExternalIdentifierAccess const &_identifierAccess,
 libyul/backends/evm/EVMCodeTransform.cpp:120:void CodeTransform::decreaseReference(YulString, Scope::Variable const &_var)
 libyul/backends/evm/EVMCodeTransform.cpp:125: unsigned &ref = m_context->variableReferences.at(&_var);
 libyul/backends/evm/EVMCodeTransform.cpp:132:bool CodeTransform::unreferenced(Scope::Variable const &_var) const
 libyul/backends/evm/EVMCodeTransform.cpp:142: for (auto const &identifier : m_scope->identifiers)
 libyul/backends/evm/EVMCodeTransform.cpp:145: Scope::Variable const &var = boost::get<Scope::Variable>(identifier.second);
 libyul/backends/evm/EVMCodeTransform.cpp:158:void CodeTransform::deleteVariable(Scope::Variable const &_var)
 libyul/backends/evm/EVMCodeTransform.cpp:168:void CodeTransform::operator()(VariableDeclaration const &_varDecl)
 libyul/backends/evm/EVMCodeTransform.cpp:190: auto &var = boost::get<Scope::Variable>(m_scope->identifiers.at(varName));
 libyul/backends/evm/EVMCodeTransform.cpp:236:void CodeTransform::operator()(Assignment const &_assignment)
 libyul/backends/evm/EVMCodeTransform.cpp:247:void CodeTransform::operator()(StackAssignment const &_assignment)
 libyul/backends/evm/EVMCodeTransform.cpp:255:void CodeTransform::operator()(ExpressionStatement const &_statement)
 libyul/backends/evm/EVMCodeTransform.cpp:262:void CodeTransform::operator()(Label const &_label)
 libyul/backends/evm/EVMCodeTransform.cpp:268: Scope::Label &label = boost::get<Scope::Label>(m_scope->identifiers.at(_label.name));
 libyul/backends/evm/EVMCodeTransform.cpp:273:void CodeTransform::operator()(FunctionCall const &_call)
 libyul/backends/evm/EVMCodeTransform.cpp:280: for (auto const &arg : _call.arguments | boost::adaptors::reversed)
 libyul/backends/evm/EVMCodeTransform.cpp:300: [&](Scope::Function &_function) { function = &_function; })),
 libyul/backends/evm/EVMCodeTransform.cpp:304: for (auto const &arg : _call.arguments | boost::adaptors::reversed)
 libyul/backends/evm/EVMCodeTransform.cpp:319:void CodeTransform::operator()(FunctionalInstruction const &_instruction)
 libyul/backends/evm/EVMCodeTransform.cpp:343: for (auto const &arg : _instruction.arguments | boost::adaptors::reversed)
 libyul/backends/evm/EVMCodeTransform.cpp:351:void CodeTransform::operator()(Identifier const &_identifier)
 libyul/backends/evm/EVMCodeTransform.cpp:357: [=](Scope::Variable &_var) {
 libyul/backends/evm/EVMCodeTransform.cpp:367: [=](Scope::Label &_label) {
 libyul/backends/evm/EVMCodeTransform.cpp:383:void CodeTransform::operator()(Literal const &_literal)
 libyul/backends/evm/EVMCodeTransform.cpp:391:void CodeTransform::operator()(yul::Instruction const &_instruction)
 libyul/backends/evm/EVMCodeTransform.cpp:401:void CodeTransform::operator()(If const &_if)
 libyul/backends/evm/EVMCodeTransform.cpp:414:void CodeTransform::operator()(Switch const &_switch)
 libyul/backends/evm/EVMCodeTransform.cpp:422: for (Case const &c : _switch.cases)
 libyul/backends/evm/EVMCodeTransform.cpp:443: for (auto const &c : caseBodies)
 libyul/backends/evm/EVMCodeTransform.cpp:462:void CodeTransform::operator()(FunctionDefinition const &_function)
 libyul/backends/evm/EVMCodeTransform.cpp:466: Scope::Function &function = boost::get<Scope::Function>(m_scope->identifiers.at(_function.name));
 libyul/backends/evm/EVMCodeTransform.cpp:473: for (auto const &v : _function.parameters | boost::adaptors::reversed)
 libyul/backends/evm/EVMCodeTransform.cpp:475: auto &var = boost::get<Scope::Variable>(varScope->identifiers.at(v.name));
 libyul/backends/evm/EVMCodeTransform.cpp:495: for (auto const &v : _function.returnVariables)
 libyul/backends/evm/EVMCodeTransform.cpp:497: auto &var = boost::get<Scope::Variable>(varScope->identifiers.at(v.name));
 libyul/backends/evm/EVMCodeTransform.cpp:517: catch (StackTooDeepError const &_error)
 libyul/backends/evm/EVMCodeTransform.cpp:583:void CodeTransform::operator()(ForLoop const &_forLoop)
 libyul/backends/evm/EVMCodeTransform.cpp:631:void CodeTransform::operator()(Break const &_break)
 libyul/backends/evm/EVMCodeTransform.cpp:636: Context::JumpInfo const &jump = m_context->forLoopStack.top().done;
 libyul/backends/evm/EVMCodeTransform.cpp:642:void CodeTransform::operator()(Continue const &_continue)
 libyul/backends/evm/EVMCodeTransform.cpp:647: Context::JumpInfo const &jump = m_context->forLoopStack.top().post;
 libyul/backends/evm/EVMCodeTransform.cpp:653:void CodeTransform::operator()(Block const &_block)
 libyul/backends/evm/EVMCodeTransform.cpp:668:AbstractAssembly::LabelID CodeTransform::labelFromIdentifier(Identifier const &_identifier)
 libyul/backends/evm/EVMCodeTransform.cpp:673: [&](Scope::Label &_label) {
 libyul/backends/evm/EVMCodeTransform.cpp:683:AbstractAssembly::LabelID CodeTransform::labelID(Scope::Label const &_label)
 libyul/backends/evm/EVMCodeTransform.cpp:690:AbstractAssembly::LabelID CodeTransform::functionEntryID(YulString _name, Scope::Function const &_function)
 libyul/backends/evm/EVMCodeTransform.cpp:701:void CodeTransform::visitExpression(Expression const &_expression)
 libyul/backends/evm/EVMCodeTransform.cpp:708:void CodeTransform::visitStatements(vector<Statement> const &_statements)
 libyul/backends/evm/EVMCodeTransform.cpp:710: for (auto const &statement : _statements)
 libyul/backends/evm/EVMCodeTransform.cpp:718:void CodeTransform::finalizeBlock(Block const &_block, int blockStartStackHeight)
 libyul/backends/evm/EVMCodeTransform.cpp:726: for (auto const &id : m_scope->identifiers)
 libyul/backends/evm/EVMCodeTransform.cpp:729: Scope::Variable const &var = boost::get<Scope::Variable>(id.second);
 libyul/backends/evm/EVMCodeTransform.cpp:745:void CodeTransform::generateMultiAssignment(vector<Identifier> const &_variableNames)
 libyul/backends/evm/EVMCodeTransform.cpp:748: for (auto const &variableName : _variableNames | boost::adaptors::reversed)
 libyul/backends/evm/EVMCodeTransform.cpp:752:void CodeTransform::generateAssignment(Identifier const &_variableName)
 libyul/backends/evm/EVMCodeTransform.cpp:757: Scope::Variable const &_var = boost::get<Scope::Variable>(*var);
 libyul/backends/evm/EVMCodeTransform.cpp:772:int CodeTransform::variableHeightDiff(Scope::Variable const &_var, YulString _varName, bool _forSwap)

```
Please check that your changes are working as intended.

ChrisChinchilla commented:
@chriseth Had an odd merge issue with files this PR didn't change, but I think this is now rebased and squashed, and the error mentioned above shouldn't trigger anymore.
